pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -1870,3 +1889,273 @@ func uplinkToPhysdev(ifname string) string {
+	status.HostIPAddrModInitialized = true
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// Assignable IP Address counter routines
+// for DhcpRange
+func setHostIPAddrCountOnDhcpRange(status *types.NetworkInstanceStatus) {
+	addr1 := status.DhcpRange.Start.To4()
+	addr2 := status.DhcpRange.End.To4()
+	if addr1 == nil || addr2 == nil {
+		return
+	}
+	start := uint(addr1[0])<<24 + uint(addr1[1])<<16 +
+		uint(addr1[2])<<8 + uint(addr1[3])
+	end := uint(addr2[0])<<24 + uint(addr2[1])<<16 +
+		uint(addr2[2])<<8 + uint(addr2[3])
+	addrcount := int(end + 1 - start)
+	status.HostIPAddressCount = getMaxIPAddrCountForNetworkInstance(addrcount)
+}
+
+// for Subnet
+func setHostIPAddrCountOnSubnet(status *types.NetworkInstanceStatus) {
+	maskLen, _ := status.Subnet.Mask.Size()
+	addrcount := 1 << (32 - maskLen)
+	// especially applicable for smaller subnets,
+	// discount the network and broadcast addresses(2)
+	// for larger subnets, it will be restricted to 254
+	// by getMaxIPAddrCountForNetworkInstance
+	addrcount -= 2
+	status.HostIPAddressCount = getMaxIPAddrCountForNetworkInstance(addrcount)
+}
+
+// for network size more than types.AppNumMax(254), restrict to
+// types.AppNumMax(254), limited by bitmap size, with a discount of 2
+// (network address and broadcast address)
+func getMaxIPAddrCountForNetworkInstance(count int) int {
+	if count > types.AppNumMax {
+		return types.AppNumMax
+	}
+	return count
+}
+
+// static ip address to app number
+func uNetAppNumForStaticGet(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appID string,
+	staticIPAddr net.IP) (int, error) {
+	if status.HostIPAddressCount == 0 {
+		return types.AppNumInvalid, nil
+	}
+	appNumOnUNet, err := persistentAddrMapAppNumGet(ctx, status.UUID.String(), appID)
+	if err == nil {
+		return appNumOnUNet, nil
+	}
+	startIP := status.DhcpRange.Start.To4()
+	if startIP == nil {
+		return types.AppNumInvalid, errors.New("parse error")
+	}
+	start := uint(startIP[0])<<24 + uint(startIP[1])<<16 +
+		uint(startIP[2])<<8 + uint(startIP[3])
+	ipAddr := staticIPAddr.To4()
+	addr := uint(ipAddr[0])<<24 + uint(ipAddr[1])<<16 +
+		uint(ipAddr[2])<<8 + uint(ipAddr[3])
+	appNumOnUNet = int(addr - start)
+	if appNumOnUNet > status.HostIPAddressCount {
+		appNumOnUNet = appNumOnUNet % status.HostIPAddressCount
+	}
+	// check whether this is already occupied
+	if status.AppNumBitMap.IsSet(appNumOnUNet) {
+		errStr := fmt.Sprintf("(%s-%s): IP Address (%s) is already occupied\n",
+			status.DisplayName, status.UUID.String(), ipAddr.String())
+		log.Errorf(errStr)
+		return types.AppNumInvalid, errors.New(errStr)
+	}
+	return appNumOnUNet, nil
+}
+
+// bridge to app number
+func uNetAppNumForBridgeGet(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) (int, error) {
+	appNumOnUNet := types.AppNumInvalid
+	// a switch variant, ignore
+	if status.Type != types.NetworkInstanceTypeLocal &&
+		status.Type != types.NetworkInstanceTypeCloud {
+		return appNumOnUNet, nil
```

Comment:
I think we won't need this function as we split the appNumOnUNet and IP allocation apart, but I note that the caller doesn't check for AppNumInvalid. So it is unknow to me what happens as we hit this condition.

Reply:
AppNumInvalid is also for switch network bridges. We are capturing the failures with Err.

Reply:
But the two callers of this function do exactly the same thing for AppNumInvalid and an error, hence there is no distiction needed and AppNumInvalid doesn't need to be defined and used.
That also fixes the error log where you log the nil error in the AppNumInvalid case.